doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
 from Julia like this is generally accomplished by creating an
 uninitialized array and passing a pointer to its data to the C function.
 
-A prefix ``&`` is used to indicate that a pointer to a scalar argument
-should be passed instead of the scalar value itself (required for all
-Fortran function arguments, as noted above). The following
-example computes a dot product using a BLAS function.
+Creating C-Compatible Julia Function Pointers
+---------------------------------------------
 
-::
+It is possible to pass Julia functions to native c-functions that accept
+function pointer arguments. For example, to match c-prototypes of the form::
 
-    function compute_dot(DX::Vector{Float64}, DY::Vector{Float64})
-      assert(length(DX) == length(DY))
-      n = length(DX)
-      incx = incy = 1
-      product = ccall( (:ddot_, "libLAPACK"),
-                      Float64,
-                      (Ptr{Int32}, Ptr{Float64}, Ptr{Int32}, Ptr{Float64}, Ptr{Int32}),
-                      &n, DX, &incx, DY, &incy)
-      return product
+    typedef returntype (*functiontype)(argumenttype,...)
+
+The function `cfunction` generates the c-compatible function pointer for
+a call to a Julia Julia library function.
+Arguments to ``cfunction`` are as follows:
+
+1. A Julia Function
+
+2. Return type
+
+3. A tuple of input types
+
+A classic example is the standard C library ``qsort`` function,
+declared as::
+
+    void qsort(void *base, size_t nmemb, size_t size,
+               int(*compare)(const void *a, const void *b));
+
+The ``base`` argument is a pointer to an array of length ``nmemb``, with elements of
+``size`` bytes each. ``compare`` is a callback function which takes pointers to two
+elements ``a`` and ``b`` and returns an integer less/greater than zero if ``a`` should
+appear before/after ``b`` (or zero if any order is permitted). Now, suppose that we
+have a 1d array ``A`` of values in Julia that we want to sort using the ``qsort``
+function (rather than Julia’s built-in ``sort`` function). Before we worry about calling
+``qsort`` and passing arguments, we need to write a comparison function that works for
+some arbitrary type T::
+
+    function mycompare{T}(a::T, b::T)
+        return convert(Cint, a < b ? -1 : a > b ? +1 : 0)::Cint
     end
 
-The meaning of prefix ``&`` is not quite the same as in C. In
-particular, any changes to the referenced variables will not be
-visible in Julia unless the type is mutable (declared via
-``type``). However, even for immutable types it will not cause any
-harm for called functions to attempt such modifications (that is,
-writing through the passed pointers). Moreover, ``&`` may be used with
-any expression, such as ``&0`` or ``&f(x)``.
+Notice that we have to be careful about the return type: ``qsort`` expects a function
+returning a C ``int``, so we must be sure to return ``Cint`` via a call to ``convert``
+and a ``typeassert``.
```

Comment:
(I wish that `cfunction` would just call `convert` and add the typeassert for you.)